### DISCUSSION TITLE
Generating Bundles without conditional references

### DISCUSSION THREAD

**Discussion by XcrigX** [synthetichealth/synthea#1332] [Q&A] [answered]
Synthea by default generates resources with conditional references to Locations, Organizations, and Practitioners. 
i.e. 

```
      "managingOrganization": [ {
        "reference": "Organization?identifier=https://github.com/synthetichealth/synthea|05cc5056-7af4-3afb-9a5a-d7a1c23a2509",
        "display": "KINDRED AT HOME"
      } ]
```

If you turn off transaction bundles it generates them as logical references instead. 
i.e.
```
      "managingOrganization": {
        "identifier": {
          "system": "https://github.com/synthetichealth/synthea",
          "value": "1b3fe590-8793-3f08-865d-01fc425a84d4"
        },
        "display": "FENWAY COMMUNITY HEALTH CENTER, INC"
      }
```

Is there a way to get it to generate them as regular literal/relative references?

**Comment by jawalonoski** [comment]
You mean like this?

```json
"managingOrganization": [ {
        "reference": "Organization/05cc5056-7af4-3afb-9a5a-d7a1c23a2509",
        "display": "KINDRED AT HOME"
      } ]
```

No, there is no configuration setting to support this at this time.

Assuming you are loading the data into a FHIR server, that would require the server allow you to set the resource ID when you create it, which is atypical behavior. See https://hl7.org/fhir/R4/http.html#create

> The request body SHALL be a FHIR Resource. The resource does not need to have an id element (this is one of the few cases where a resource exists without an id element). If an id is provided, the server SHALL ignore it. ... The server SHALL populate the id, meta.versionId and meta.lastUpdated with the new correct values.

In the transaction, the server should rewrite those conditional references to actual identifiers, as described here https://hl7.org/fhir/R4/http.html#trules.

If you have a different use-case that needs references of this variety, the FHIR exporters would require modification.

**Reply by XcrigX** [reply]
Thanks @jawalonoski - yes, that is what I mean. For instance the old SyntheticMass 1k bundles are self-contained with no conditional references. I understand why you are doing it that way, and that if the destination is a FHIR server that the references will be resolved with the transaction (assuming you preloaded the matching data). 

Sometimes I use Synthea for test data where the end goal is not a FHIR server, however. For instance, testing a FHIR to X converter - where having fully self-contained bundles with non-conditional/logical references would be useful.

**Reply by jawalonoski** [reply, marked as answer]
Created issue https://github.com/synthetichealth/synthea/issues/1333

**Reply by jawalonoski** [reply]
> Sometimes I use Synthea for test data where the end goal is not a FHIR server, however. For instance, testing a FHIR to X converter - where having fully self-contained bundles with non-conditional/logical references would be useful.

I assume "fully self-contained bundle" does not mean the shared resources are repeated in each Bundle, only that the references are resolved to the identifiers we've assigned.

**Reply by XcrigX** [reply]
I was thinking the shared resources would have to be repeated in each Bundle (with a conditional create). I can already set _exporter.fhir.transaction_bundle=false_ to get it to generate a logical reference to the target resource identifier, if that's what you mean.

**Comment by XcrigX** [comment]
Otherwise you'd have to use PUT operations in the supplemental hospitalinformation/practitionerinformation bundles so the resource IDs would match if they were applied to a FHIR server.. I think.

**Comment by XcrigX** [comment]
Churning it in my head a bit, maybe that's the most general/generic option that could be useful for other use cases and still be fully FHIR compliant. 
Have an option like:  generateSelfContainedBundles or some such that just includes the shared resources that are referenced in any given Bundle, within the Bundle. 
Perhaps could be useful for systems that are not capable of processing conditional references, people wanting to simulate message bundles that would be fully self-contained coming from some external entity, people generating bundles for unit tests, etc.

**Comment by dmitry-zasypkin** [comment]
When producing .ndjson files, Synthea generates conditional references to Locations, Organizations, and Practitioners. Searching through the FHIR specs, it seems that conditional references are only allowed within Bundles. Adding an option to generate literal references instead of conditional references would allow to produce valid .ndjson files.
